### PR TITLE
Fix/v2 pda seed limits

### PR DESCRIPTION
**Pull request by 0x4ka5h**
#### Finding
`lang-v2` PDA helpers accepted seed counts and seed lengths that were looser than Solana runtime limits. That could allow derivations that compile locally but fail once the runtime enforces the real bounds.

#### Fix
This PR enforces the runtime PDA limits across the helper surface: each seed is capped at 32 bytes and total seed counts now match Solana’s limits, including bump-aware paths. It also adds boundary tests covering accepted maximum sizes and rejected oversized or overcounted seeds.

replaces https://github.com/otter-sec/anchor/pull/4813

### CODE REVIEW

File: lang-v2/derive/src/pda.rs [hunk 1]
```diff
@@ -17,6 +17,14 @@ use {
...
 
 const PDA_MARKER: &[u8; 21] = b"ProgramDerivedAddress";
+const MAX_PDA_SEEDS_TOTAL: usize = 16;
+const MAX_PDA_SEED_LEN: usize = 32;
+const MAX_PDA_SEEDS_WITHOUT_BUMP: usize = MAX_PDA_SEEDS_TOTAL - 1;
+
+const _: () = {
+    assert!(MAX_PDA_SEEDS_TOTAL == solana_address::MAX_SEEDS);
+    assert!(MAX_PDA_SEED_LEN == solana_address::MAX_SEED_LEN);
+};
```
Comment (swaroop-osec) [inline lang-v2/derive/src/pda.rs]: Since we've added the `solana-address` dependency, we don't need this assertion anymore. We can just import it directly from the crate.

```suggestion
const MAX_PDA_SEEDS_TOTAL: usize = solana_address::MAX_SEEDS;
const MAX_PDA_SEED_LEN: usize =  solana_address::MAX_SEED_LEN;
const MAX_PDA_SEEDS_WITHOUT_BUMP: usize = MAX_PDA_SEEDS_TOTAL - 1;
```

File: lang-v2/src/cpi.rs [hunk 2]
```diff
@@ -38,6 +45,14 @@ fn encode_system_transfer(lamports: u64) -> [u8; 12] {
     data
 }
 
+#[inline(always)]
+fn validate_pda_seeds(seeds: &[&[u8]], max_seed_count: usize) -> Result<(), ProgramError> {
+    if seeds.len() > max_seed_count || seeds.iter().any(|seed| seed.len() > MAX_PDA_SEED_LEN) {
+        return Err(ProgramError::InvalidSeeds);
```
Comment (swaroop-osec) [inline lang-v2/src/cpi.rs]:
```suggestion
        return Err(ProgramError::MaxSeedLengthExceeded);
```
Match Solana/Pinocchio error

File: lang-v2/src/cpi.rs [hunk 3]
```diff
@@ -393,6 +417,69 @@ fn hash_pda_seeds(seeds: &[&[u8]], program_id: &Address) -> Result<Address, Prog
     Ok(Address::new_from_array(unsafe { hash.assume_init() }))
 }
 
+#[cfg(test)]
```
Comment (swaroop-osec) [inline lang-v2/src/cpi.rs:407]: We need to update tests here with same `ProgramError::MaxSeedLengthExceeded`

**Review by swaroop-osec — CHANGES_REQUESTED** [review]

File: lang-v2/src/cpi.rs (hunk 3 above)
Comment (akash-osec) [reply lang-v2/src/cpi.rs:407]: you meant only in cfg tests or everywhere `ProgramError::InvalidSeeds` is used?

Comment (swaroop-osec) [reply lang-v2/src/cpi.rs:407]: I mean if we change the error these assets will fail so we need to update them here in the tests.

Comment (akash-osec) [reply lang-v2/src/cpi.rs:407]: updated. please check.

**Comment by swaroop-osec** [comment]
@0x4ka5h Address the failed tests

**Review by swaroop-osec — APPROVED** [review]
